### PR TITLE
Add `/api/speech` endpoint

### LINKED ISSUE #501: Speech from UI does not work when using RBAC
# Describe the bug
The UI fetches the key for the Azure Speech Service from the `/api/config` endpoint, however there is no value for the key when using RBAC, therefore it fails. We need to somehow get speech service to work even when there is no key.

# Expected behavior
Can input query from the UI using voice.

# How does this bug make you feel?
😖

---

# Debugging information

## Steps to reproduce
* Deploy solution with `AZURE_AUTH_TYPE=rbac`
* Navigate to UI
* Press microphone
* Check console

## Screenshots

image

---

# Tasks

_To be filled in by the engineer picking up the issue_

- [ ] Task 1
- [ ] Task 2
- [ ] ...

### PR DESCRIPTION
**Pull request by cecheta**
Closes #501, required by #101 

## Purpose

This PR adds a new `/api/speech` endpoint, which uses the Speech Key to return a short-lived token that the frontend can use to call the speech service. 
The frontend calls the endpoint every time the microphone button is pressed.
When using RBAC, the speech key is fetched directly from Azure as it is not present on any env var.

https://github.com/Azure-Samples/cognitive-services-speech-sdk/blob/master/samples/js/browser/README.md#token-exchange-process

The PR also refactors some of the tests/mocking in `test_app.py` ,so the mocking of the env_helper is done in one place.

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
* Deploy application with Keys + with RBAC

## What to Check
Verify that the following are valid
* Test that speech-to-text is working

### CODE REVIEW

File: infra/main.bicep [hunk 1]
```diff
@@ -398,6 +398,7 @@ module web './app/web.bicep' = if (hostingModel == 'code') {
       AZURE_OPENAI_API_VERSION: azureOpenAIApiVersion
       AZURE_OPENAI_STREAM: azureOpenAIStream
       AZURE_OPENAI_EMBEDDING_MODEL: azureOpenAIEmbeddingModel
+      AZURE_RESOURCE_GROUP: rgName
```
Comment (ross-p-smith) [inline infra/main.bicep:401]: If we are going to make this change to every app service - we could move this down to `/infra/core/host/appservice.bicep`

**Comment by ross-p-smith** [comment]
I think you've just made the same comment - but maybe some co-ordination with #660. It feels like maybe this goes in first and shows how to get config? What are your thoughts?

**Comment by cecheta** [comment]
> I think you've just made the same comment - but maybe some co-ordination with #660. It feels like maybe this goes in first and shows how to get config? What are your thoughts?

Actually, this PR removes the ability to fetch config from the `/api/config` endpoint, but now I'm thinking it would actually be better to leave that in, so it can be used in #660 

Will update

**Comment by superhindupur** [comment]
@cecheta so we're on the same page:
- I'm now convinced that injecting the env variable at buildtime into the frontend typescript app in https://github.com/Azure-Samples/chat-with-your-data-solution-accelerator/pull/660 is overly complicating the build-time dependencies. 
- I agree that /api/config is a good place to add the speech languages 
- Should I go ahead and remove the azureSpeechKey from the /api/config endpoint in my PR so that we don't return the key from there?

File: infra/main.bicep (hunk 1 above)
Comment (cecheta) [reply infra/main.bicep]: Done

File: infra/main.bicep [hunk 2]
```diff
@@ -877,8 +878,10 @@ output AZURE_SEARCH_TITLE_COLUMN string = azureSearchTitleColumn
 output AZURE_SEARCH_URL_COLUMN string = azureSearchUrlColumn
 output AZURE_SEARCH_USE_INTEGRATED_VECTORIZATION bool = azureSearchUseIntegratedVectorization
 output AZURE_SEARCH_INDEX string = azureSearchIndex
+output AZURE_SPEECH_SERVICE_NAME string = speechServiceName
 output AZURE_SPEECH_SERVICE_REGION string = location
 output AZURE_SPEECH_SERVICE_KEY string = useKeyVault ? storekeys.outputs.SPEECH_KEY_NAME : ''
+output AZURE_SUBSCRIPTION_ID string = subscription().subscriptionId
```
Comment (ross-p-smith) [inline infra/main.bicep:884]: is this already output by azd? Suppose doesnt hurt if we are explicit - I'll let you close this based on your opinion

File: code/create_app.py [hunk 3]
```diff
@@ -285,6 +287,24 @@ def conversation_without_data(request, env_helper):
...
+def get_speech_key(env_helper: EnvHelper):
+    """
+    Get the Azure Speech key directly from Azure.
+    This is required to generate short-lived tokens when using RBAC.
+    """
+    client = CognitiveServicesManagementClient(
+        credential=DefaultAzureCredential(),
+        subscription_id=env_helper.AZURE_SUBSCRIPTION_ID,
+    )
+    keys = client.accounts.list_keys(
```
Comment (adamdougal) [inline code/create_app.py:300]: Does listing keys even work with the RBAC auth type?

File: code/create_app.py [hunk 4]
```diff
@@ -285,6 +287,24 @@ def conversation_without_data(request, env_helper):
         )
 
 
+@functools.cache
```
Comment (ross-p-smith) [inline code/create_app.py:290]: Does this file get used when as a function? I'm going based on the dockerfile and I dont think it copies this in

File: code/frontend/src/pages/chat/Chat.tsx [hunk 5]
```diff
@@ -155,41 +147,47 @@ const Chat = () => {
...
 
-  const startSpeechRecognition = (
-    subscriptionKey: string,
-    serviceRegion: string
-  ) => {
+  const fetchSpeechToken = async (): Promise<{ token: string, region: string; }> => {
+    try {
+      const response = await fetch("/api/speech/token");
+      if (!response.ok) {
+        throw new Error("Network response was not ok");
```
Comment (adamdougal) [inline code/frontend/src/pages/chat/Chat.tsx:154]: Should we log out the response to aid future debugging?

File: code/tests/functional/backend_api/tests/with_data/test_speech_token.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+from tests.functional.backend_api.app_config import AppConfig
+from tests.functional.backend_api.request_matching import (
+    RequestMatcher,
+    verify_request_made,
+)
+
+pytestmark = pytest.mark.functional
+
+
+def test_speech_token_returned(app_url: str, app_config: AppConfig):
```
Comment (adamdougal) [inline code/tests/functional/backend_api/tests/with_data/test_speech_token.py:13]: 👍

File: infra/main.bicep (hunk 2 above)
Comment (cecheta) [reply infra/main.bicep]: Good point, removed

File: code/create_app.py (hunk 4 above)
Comment (cecheta) [reply code/create_app.py:290]: Would the function need this?

File: code/create_app.py (hunk 3 above)
Comment (cecheta) [reply code/create_app.py:300]: Yes it does, we've already assigned the Cognitive Services User role so it has permissions

File: code/frontend/src/pages/chat/Chat.tsx (hunk 5 above)
Comment (cecheta) [reply code/frontend/src/pages/chat/Chat.tsx:155]: Added

**Comment by superhindupur** [comment]
@cecheta and I chatted offline - the conclusion was that we will rename the /api/config to /api/speech and return the token from there. Then I will add the languages to it under #660

**Review by adamdougal — APPROVED** [review]
noice

**Review by ross-p-smith — DISMISSED** [review]
😎

File: code/frontend/src/pages/chat/Chat.tsx [hunk 7]
```diff
@@ -134,62 +131,48 @@ const Chat = () => {
...
-        );
-      } else {
-        const recognizer = multiLingualSpeechRecognizer(subscriptionKey, serviceRegion, [
-          "en-US",
-          "fr-FR",
-          "de-DE",
-          "it-IT"
-        ]);
-        recognizerRef.current = recognizer; // Store the recognizer in the ref
+      const { token, region } = await fetchSpeechToken();
```
Comment (superhindupur) [inline code/frontend/src/pages/chat/Chat.tsx:152]: Hey does this mean that the config endpoint will be called every time the microphone is clicked?

**Review by superhindupur — APPROVED** [review]

**Review by ross-p-smith — APPROVED** [review]

**Review by adamdougal — APPROVED** [review]

**Pull request merged**